xclim/indices/run_length.py

```diff
@@ -88,83 +88,30 @@ def rle(
-    i = xr.DataArray(np.arange(da[dim].size), dims=dim)
-    if use_dask:
-        i = i.chunk({dim: -1})
-
-    ind, da = xr.broadcast(i, da)
-    if use_dask:
-        # Rechunk, but with broadcasted da
-        ind = ind.chunk(da.chunks)
-
-    b = ind.where(~da)  # find indexes where false
-    end1 = (
-        da.where(b[dim] == b[dim][-1], drop=True) * 0 + n
-    )  # add additional end value index (deal with end cases)
-    start1 = (
-        da.where(b[dim] == b[dim][0], drop=True) * 0 - 1
-    )  # add additional start index (deal with end cases)
-    b = xr.concat([start1, b, end1], dim)
-
-    # Ensure bfill operates on entire (unchunked) time dimension
-    # Determine appropriate chunk size for other dims, follow logic from dask:
-    #   do not exceed 'array.chunk-size' unless array.slicing.split-large-chunks is False.
-    # Skip this if there's already only one chunk along dim
-    ndims = len(b.shape)
-    if (
-        use_dask
-        and len(b.chunks[b.get_axis_num(dim)]) > 1
-        and dask.config.get("array.slicing.split-large-chunks") is not False
-        and np.prod(da.data.chunksize) > max_chunk
-    ):
-        chunk_dim = b[dim].size
-        # divide extra dims into equal size
-        # Note : even if calculated chunksize > dim.size result will have chunk==dim.size
-        chunksize_ex_dims = None  # TODO: This raises type assignment errors in mypy
-        if ndims > 1:
-            if dask.config.get("array.slicing.split-large-chunks") is None:
-                warn(
-                    (
-                        f"xclim's run length requires a single chunk along {dim} "
-                        "and must rechunk the provided array. To control this behaviour, "
-                        "use the dask.config entries for `array.chunk-size` and "
-                        "`array.slicing.split-large-chunks`."
-                    ),
-                    category=dask.array.PerformanceWarning,
-                )
-            chunksize_ex_dims = np.round(
-                np.power(max_chunk / chunk_dim, 1 / (ndims - 1))
-            )
-        chunks = dict()
-        chunks[dim] = -1
-        for dd in b.dims:
-            if dd != dim:
-                chunks[dd] = chunksize_ex_dims
-        b = b.chunk(chunks)
-
-    # back-fill nans with first position after
-    z = b.bfill(dim=dim)
-
-    # calculate lengths
-    d = z.diff(dim=dim) - 1
-    d = d.where(d >= 0)
-    d = d.isel({dim: slice(None, -1)}).where(da, 0)
-    if index == "last":
-        d = d.reindex({dim: d[dim][::-1]})
-    return d
+    da = da.astype(int)
+
+    # "first" case: Algorithm is applied on inverted array and output is inverted back
+    if index == "first":
+        da = da[{dim: slice(None, None, -1)}]
+
+    # Get cumulative sum for each series of 1, e.g. da == 100110111 -> cs_s == 100120123
+    cs = da.cumsum(dim=dim)
+    cs2 = cs.where(da == 0)
+    cs2[{dim: 0}] = 0
+    cs2 = cs2.ffill(dim=dim)
+    cs_s = cs - cs2
```

Comment:
I feel that these few lines of code are really great !
They should probably be in a separate function because this is basically a cumsum with a reset mechanism when encountering a zero so I think it can be useful in a few other cases.
Also it would help for readability.

Reply:
Ok sounds good! I guess `core/utils` would be the right place?  Let me know if you think there's a better place.

Reply:
Yes, it's just an intuition because I would expect other cases where a `cumsum_with_reset` would be useful.

I would make it a private function in run_length module, mainly for readability, until it needs to be reused elsewhere. But that's just my way of doing things.